### PR TITLE
[RFC] vim-patch: the completion patches

### LINKED ISSUE #972: VIM 7.4: Possible regression via patch 2146?
Hello,

With patch 2146 (472472898ab71ac80a86fedc37f8eb91461788dd), text may be inserted while backspacing or typing if you have configured VIM to automatically display a popup menu with possible completions.  This is most noticeable when trying to backspace on a word that has a possible match.

Using `vim -u NONE` alleviates this use, but that would be expected.

I'm currently using [ACP](http://www.vim.org/scripts/script.php?script_id=1879) to handle autocompletion.

Using the below patch fixes/reverts the issue (on HEAD) for me:

<pre>diff --git a/src/edit.c b/src/edit.c
index 6dae4a0..2281232 100644
--- a/src/edit.c
+++ b/src/edit.c
@@ -3893,7 +3893,7 @@ ins_compl_prep(int c)

            /* CTRL-E means completion is Ended, go back to the typed text.
             * but only do this, if the Popup is still visible */
-           if (c == Ctrl_E && pum_visible())
+           if (c == Ctrl_E)
            {
                ins_compl_delete();

</pre>

Hopefully someone can shed some insight on this.  I'm unsure if it's a bug in VIM or the plugins implementation of the VIM popup via ACP.

Below is the git bisect where I traced from patch 1051(af8af8bfac5792fa64efbc524032d568cc7754f7) to patch 2181(a772ec5e91efbd5c4b99bb5cf5704a6c6f0baf3f):

<pre>git bisect start
# good: [af8af8bfac5792fa64efbc524032d568cc7754f7] patch 7.4.1051 Problem:    Segfault when unletting "count". Solution:   Check for readonly and locked first. (Dominique Pelle)             Add a test.
git bisect good af8af8bfac5792fa64efbc524032d568cc7754f7
# bad: [a772ec5e91efbd5c4b99bb5cf5704a6c6f0baf3f] patch 7.4.2181 Problem:    Compiler warning for unused variable. Solution:   Remove it. (Dominique Pelle)
git bisect bad a772ec5e91efbd5c4b99bb5cf5704a6c6f0baf3f
# good: [a4f6ca717b4483eb82c6c71f71a5a5cf70e55d80] patch 7.4.1618 Problem:    Starting job with output to buffer changes options in the current             buffer. Solution:   Set "curbuf" earlier. (Yasuhiro Matsumoto)
git bisect good a4f6ca717b4483eb82c6c71f71a5a5cf70e55d80
# good: [28b942a064dd486cc241894b625ab72f5a5c6d1b] patch 7.4.1900 Problem:    Using CTRL-] in the help on "{address}." doesn't work. Solution:   Recognize an item in {}. (Hirohito Higashi, closes #814)
git bisect good 28b942a064dd486cc241894b625ab72f5a5c6d1b
# good: [93431df9eb02f7cf3d7f2142bb1bef24c5f325b2] patch 7.4.2043 Problem:    setbuvfar() causes a screen redraw. Solution:   Only use aucmd_prepbuf() for options.
git bisect good 93431df9eb02f7cf3d7f2142bb1bef24c5f325b2
# good: [ad9c2a08f0509294269a2f11a59a438b944bdd5a] patch 7.4.2109 Problem:    Setting 'display' to "lastline" is a drastic change, while             omitting it results in lots of "@" lines. Solution:   Add "truncate" to show "@@@" for a truncated line.
git bisect good ad9c2a08f0509294269a2f11a59a438b944bdd5a
# good: [bc8801c9317eb721a2ee91322669f2dd5d136380] Updated runtime files.
git bisect good bc8801c9317eb721a2ee91322669f2dd5d136380
# bad: [7522f6982197f83a5c0f6e9af07fb713934f824a] patch 7.4.2162 Problem:    Result of getcompletion('', 'sign') depends on previous             completion. Solution:   Call set_context_in_sign_cmd(). (Dominique Pelle)
git bisect bad 7522f6982197f83a5c0f6e9af07fb713934f824a
# bad: [ee695f787ade7fd88fc5f5497553d95c0c3645b5] patch 7.4.2152 Problem:    No proper translation of messages with a count. Solution:   Use ngettext(). (Sergey Alyoshin)
git bisect bad ee695f787ade7fd88fc5f5497553d95c0c3645b5
# bad: [abb71fbd399772d467aaa7b34b958b0f975c7e65] patch 7.4.2147 Problem:    test_alot fails. Solution:   Close window.
git bisect bad abb71fbd399772d467aaa7b34b958b0f975c7e65
# good: [86f2cd5bc574c23fa276d7f57cd1300e24222913] patch 7.4.2145 Problem:    Win32: Using CreateThread/ExitThread is not safe. Solution:   Use _beginthreadex and return from the thread. (Ken Takata)
git bisect good 86f2cd5bc574c23fa276d7f57cd1300e24222913
# bad: [472472898ab71ac80a86fedc37f8eb91461788dd] patch 7.4.2146 Problem:    Not enough testing for popup menu.  CTRL-E does not always work             properly. Solution:   Add more tests.  When using CTRL-E check if the popup menu is             visible. (Christian Brabandt)
git bisect bad 472472898ab71ac80a86fedc37f8eb91461788dd
# first bad commit: [472472898ab71ac80a86fedc37f8eb91461788dd] patch 7.4.2146 Problem:    Not enough testing for popup menu.  CTRL-E does not always work             properly. Solution:   Add more tests.  When using CTRL-E check if the popup menu is             visible. (Christian Brabandt)</pre>

### LINKED ISSUE #1200: Vim hangs and segfaults with a scratch PHP buffer
I can cause erratic behaviour using terminal Vim 8.0p46 on macOS 10.12.1, as follows:

```
vim -N -u NONE
filetype on
filetype plugin on
vnew +setlocal\ buftype=nofile\ bufhidden=wipe\ noswapfile
set ft=php
```

Then start typing: `i<?php<cr>for<c-x><c-o>` (the first `<` is a literal less-than character). If that is not enough, try typing `<c-n>`, `<c-p>`, and/or `<tab>` a few times, and if that is not enough yet, clear the buffer, `set modifiable` if necessary (for some reason, sometimes the buffer becomes unmodifiable) and repeat. The crash does not seem deterministic, so you may have to try a few times.

Vim sometimes just freezes and some other times hangs with `Vim: Caught deadly signal SEGV`. In both cases, the CPU usage skyrockets and I have to kill the process myself.

Note that I have recompiled Vim after upgrading to macOS 10.12.1, to be sure that the bug does not depend on some library updates in macOS.

Vim version:

```
VIM - Vi IMproved 8.0 (2016 Sep 12, compiled Oct 26 2016 21:07:47)
MacOS X (unix) version
Included patches: 1-46
Compiled by Homebrew
Huge version without GUI.  Features included (+) or not (-):
+acl             +conceal         +file_in_path    +libcall         +mouse_sgr       +profile         +tag_old_static  +viminfo
+arabic          +cryptv          +find_in_path    +linebreak       -mouse_sysmouse  +python          -tag_any_white   +vreplace
+autocmd         +cscope          +float           +lispindent      +mouse_urxvt     -python3         -tcl             +wildignore
-balloon_eval    +cursorbind      +folding         +listcmds        +mouse_xterm     +quickfix        +termguicolors   +wildmenu
-browse          +cursorshape     -footer          +localmap        +multi_byte      +reltime         +terminfo        +windows
++builtin_terms  +dialog_con      +fork()          +lua             +multi_lang      +rightleft       +termresponse    +writebackup
+byte_offset     +diff            -gettext         +menu            -mzscheme        +ruby            +textobjects     -X11
+channel         +digraphs        -hangul_input    +mksession       +netbeans_intg   +scrollbind      +timers          -xfontset
+cindent         -dnd             +iconv           +modify_fname    +num64           +signs           +title           -xim
-clientserver    -ebcdic          +insert_expand   +mouse           +packages        +smartindent     -toolbar         -xpm
+clipboard       +emacs_tags      +job             -mouseshape      +path_extra      +startuptime     +user_commands   -xsmp
+cmdline_compl   +eval            +jumplist        +mouse_dec       +perl            +statusline      +vertsplit       -xterm_clipboard
+cmdline_hist    +ex_extra        +keymap          -mouse_gpm       +persistent_undo -sun_workshop    +virtualedit     -xterm_save
+cmdline_info    +extra_search    +lambda          -mouse_jsbterm   +postscript      +syntax          +visual
+comments        +farsi           +langmap         +mouse_netterm   +printer         +tag_binary      +visualextra
   system vimrc file: "$VIM/vimrc"
     user vimrc file: "$HOME/.vimrc"
 2nd user vimrc file: "~/.vim/vimrc"
      user exrc file: "$HOME/.exrc"
       defaults file: "$VIMRUNTIME/defaults.vim"
  fall-back for $VIM: "/usr/local/share/vim"
Compilation: clang -c -I. -Iproto -DHAVE_CONFIG_H   -DMACOS_X_UNIX  -g -O2 -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1
Linking: clang   -L. -L/usr/local/lib  -L/usr/local/lib -o vim        -lm  -lncurses -liconv -framework Cocoa  -pagezero_size 10000 -image_ba
se 100000000 -L/usr/local/lib -lluajit-5.1 -fstack-protector  -L/System/Library/Perl/5.18/darwin-thread-multi-2level/CORE -lperl -F/usr/local
/opt/python/Frameworks -framework Python   -lruby.2.0.0 -lobjc
```

### LINKED ISSUE #1241: Popup menu does not default to below the cursor
A [recent patch](https://github.com/vim/vim/commit/91e44a3305ef6bf2d43496c351dcff0a45c6bfb8#diff-d21807d1e6c9ea7784b05e18791acca5R106) introduced a bug that removed `pum_height` from the consideration of where to put the popup menu.  After that patch, the popup menu always shows up above the cursor when the cursor is in the bottom half of the screen, even when there is plenty of space below the cursor for the popup menu to be displayed.  This behavior is pretty obviously seen in the line of code that I linked, and pretty obviously not desirable, based on the comment directly above that line.

### LINKED ISSUE #1312: Exiting completion menu results in unexpected change
Hello,

If I write the following in `/tmp/vimrc.vim`:
```
setlocal tw=78 cc=78
"                                                        zzz
" zzzyyyyyyyyyyyyyyyyyyy
```

Then I launch Vim from the shell like this:

    vim -Nu /tmp/vimrc.vim /tmp/vimrc.vim

Finally, in insert mode I position the cursor after `zzz` on the second line, hit `<C-x><C-n>` to open the completion menu in which the only suggestion is `zzzyyyyyyyyyyyyyyyyyyy`, hit `<C-e>` to exit, here's the result:

```
setlocal tw=78 cc=78
"
zzzyyyyyyyyyyyyyyyyyyyzzz
" zzzyyyyyyyyyyyyyyyyyyy
```

I expected the buffer to not change since I exited the menu. Besides `zzz` is inserted twice.

I'm using Vim version 8.0 (patches 1-124 included) in linux, and xfce4-terminal.

### PR DESCRIPTION
**Pull request by Shougo**
vim-patch:7.4.2146

Problem:    Not enough testing for popup menu.  CTRL-E does not always work
            properly.
Solution:   Add more tests.  When using CTRL-E check if the popup menu is
            visible. (Christian Brabandt)

vim/vim@4724728

vim-patch:7.4.2147

Problem:    test_alot fails.
Solution:   Close window.

vim/vim@abb71fb

vim-patch:7.4.2149

Problem:    If a test leaves a window open a following test may fail.
Solution:   Always close extra windows after running a test.

vim/vim@7cba71d

vim-patch:7.4.2188

Problem:    Completion does not work properly with some plugins.
Solution:   Revert the part related to typing CTRL-E. (closes vim/vim#972)

vim/vim@c9fb77c

vim-patch:7.4.2321

Problem:    When a test is commented out we forget about it.
Solution:   Let a test throw an exception with "Skipped" and list skipped test
            functions. (Christian Brabandt)

https://github.com/vim/vim/commit/dac1947bb366ef43cd6da95acc730554e76d8b84

vim-patch:7.4.2331

Problem:    Using CTRL-X CTRL-V to complete a command line from Insert mode
            does not work after entering an expression on the command line.
Solution:   Don't use "ccline" when not actually using a command line. (test
            by Hirohito Higashi)

https://github.com/vim/vim/commit/33a80eeb859a78ba93432da6fa585786cfd77249

vim-patch:8.0.0008

Problem:    Popup complete test is disabled.
Solution:   Enable the test and change the assert. (Hirohito Higashi)

https://github.com/vim/vim/commit/9e02cfa226b2577ec867b544a1a450a428a19880

vim-patch:8.0.0047

Problem:    Crash when using the preview window from an unnamed buffer.
            (lifepillar)
Solution:   Do not clear the wrong buffer. (closes vim/vim#1200)

https://github.com/vim/vim/commit/50e5376926dc2ec4a26a7a16f8f0f3213c4afdf0

vim-patch:8.0.0053

Problem:    No test for what 8.0.0047 fixes.
Solution:   Add a test. (Hirohito Higashi)

https://github.com/vim/vim/commit/60ef3e81f4a54d9f7ee617d57021f0811ec8ada5

vim-patch:8.0.0058

Problem:    Positioning of the popup menu is not good.
Solution:   Position it better. (Hirohito Higashi)

vim/vim@91e44a3

vim-patch:8.0.0099

Problem:    Popup menu always appears above the cursor when it is in the lower
            half of the screen. (Matt Gardner)
Solution:   Compute the available space better. (Hirohito Higashi,
            closes vim/vim#1241)

vim/vim@7309528

vim-patch:8.0.0127

Problem:    Cancelling completion still inserts text when formatting is done
            for 'textwidth'. (lacygoill)
Solution:   Don't format when CTRL-E was typed. (Hirohito Higashi,
            closes vim/vim#1312)

https://github.com/vim/vim/commit/73fd4988866c3adc15b5d093efdf5e8cf70d093d

### CODE REVIEW

**Comment by Shougo** [comment]
Missing patches:

7.4.2146, 2147, 2149

**Comment by Shougo** [comment]
I have fixed the build error.

**Comment by Shougo** [comment]
To fix the functional tests error, Vim 7.4.2321,7.4.2331, 8.0008, 8.0047, 8.0053 patches are needed.

**Comment by Shougo** [comment]
The fix is done.

**Comment by Shougo** [comment]
The lint errors are fixed.

**Comment by Shougo** [comment]
The all tests are passed!

File: src/nvim/popupmnu.c [hunk 1]
```diff
@@ -200,12 +201,10 @@ void pum_display(pumitem_T *array, int size, int selected, bool array_changed)
...
   }
 
-  // If there is a preview window at the top avoid drawing over it.
-  if (firstwin->w_p_pvw
-      && (pum_row < firstwin->w_height)
-      && (pum_height > firstwin->w_height + 4)) {
-    pum_row += firstwin->w_height;
-    pum_height -= firstwin->w_height;
+  // If there is a preview window at the above avoid drawing over it.
+  if (pum_row < above_row && pum_height > above_row) {
```
Comment (justinmk) [inline src/nvim/popupmnu.c]: the original patch checks for NULL:

    if (pvwin != NULL && pum_row < above_row && pum_height > above_row)

File: src/nvim/popupmnu.c [hunk 2]
```diff
@@ -123,17 +125,14 @@ void pum_display(pumitem_T *array, int size, int selected, bool array_changed)
...
-  }
-
-  // When the preview window is at the bottom stop just above it.  Also
-  // avoid drawing over the status line so that it's clear there is a window
-  // boundary.
-  if (lastwin->w_p_pvw) {
-    above_row -= lastwin->w_height + lastwin->w_status_height + 1;
+  FOR_ALL_WINDOWS_IN_TAB(pvwin, curtab) {
+      if (pvwin->w_p_pvw) {
+          if (pvwin->w_wrow < curwin->w_wrow) {
```
Comment (justinmk) [inline src/nvim/popupmnu.c]: this block is outside of the `FOR_ALL_...` block in Vim.

Comment (justinmk) [inline src/nvim/popupmnu.c]: I'll fix this before merging.

**Comment by justinmk** [comment]
Merged, thanks @Shougo.

- Made the changes mentioned above. 
- Verified some tests.

Given that some commits were out of sync with the original patches, but the final result is sound, I squashed the commits into a single commit.

**Comment by jamessan** [comment]
This missed the runtest.vim changes of 7.4.2149...  I'll fix that in #5613, since the tests are failing there because of that.

**Comment by Shougo** [comment]
Thank you!

**Pull request closed without merging**